### PR TITLE
Rewrite most built-in Array methods as NativeFunctions

### PR DESCRIPTION
**Pull request by cpcallen**
Only `Array.prototype.toString` and `Array.prototype.join` still use their old implementation; they will be dealt with in the next PR.

### CODE REVIEW

File: server/interpreter.js [hunk 1]
```diff
@@ -1098,6 +1090,52 @@ Interpreter.prototype.initArray_ = function() {
...
+    call:  function(intrp, thread, state, thisVal, args) {
+      var searchElement = args[0];
+      var fromIndex = args[1];
+      var perms = state.scope.perms;
+      var obj = intrp.toObject(thisVal, perms);
+      var len = Interpreter.toLength(obj.get('length', perms));
+      if (len === 0) return -1;
+      var n = (fromIndex === undefined ? 0 : Interpreter.toInteger(fromIndex));
+      if (n >= len) return -1;
+      var k = (n >= 0 ? n : Math.max(len - Math.abs(n), 0));
```
Comment (NeilFraser) [inline server/interpreter.js]: Move the parens to just around (n >= 0)

File: server/interpreter.js [hunk 2]
```diff
@@ -1098,6 +1090,52 @@ Interpreter.prototype.initArray_ = function() {
...
+    call:  function(intrp, thread, state, thisVal, args) {
+      var searchElement = args[0];
+      var fromIndex = args[1];
+      var perms = state.scope.perms;
+      var obj = intrp.toObject(thisVal, perms);
+      var len = Interpreter.toLength(obj.get('length', perms));
+      if (len === 0) return -1;
+      var n = (fromIndex === undefined ? len - 1 :
+          Interpreter.toInteger(fromIndex));
+      var k = (n >= 0 ? Math.min(n, len - 1) : len - Math.abs(n));
```
Comment (NeilFraser) [inline server/interpreter.js]: Move parens on these two lines too.

File: server/interpreter.js [hunk 3]
```diff
@@ -1881,6 +1881,48 @@ Interpreter.legalArrayIndex = function(x) {
...
+  } else if (number === 0 || number === Infinity || number === -Infinity) {
+    return number;
+  }
+  return Math.trunc(number);
+};
+
+/**
+ * The ToUint32 function from ES6 §7.1.6.  The abstract operation
+ * ToUint32 converts argument to one of 2**32 integer values in the
+ * range 0 through 2**32−1, inclusive.
```
[the remark this answers is not in the document]
Comment (cpcallen) [reply server/interpreter.js]: All `^` denoting exponentiation replaced by `**`.

File: server/interpreter.js (hunk 1 above)
Comment (cpcallen) [reply server/interpreter.js]: Done.

File: server/interpreter.js (hunk 2 above)
Comment (cpcallen) [reply server/interpreter.js]: Done (and elsewhere).

**Pull request merged**